horovod/run/common/util/safe_shell_exec.py

```diff
@@ -42,24 +44,32 @@ def terminate_executor_shell_and_children(pid):
             pass
 
     # Wait for graceful termination.
-    time.sleep(GRACEFUL_TERMINATION_TIME_S)
+    gone, alive = psutil.wait_procs(p.children(), timeout=GRACEFUL_TERMINATION_TIME_S)
 
     # Send STOP to executor shell to stop progress.
     p.send_signal(signal.SIGSTOP)
 
     # Kill children recursively.
-    for child in p.children(recursive=True):
+    for child in alive:
         try:
+            for grandchild in child.children(recursive=True):
+                try:
+                    grandchild.kill()
+                except psutil.NoSuchProcess:
+                    pass
             child.kill()
         except psutil.NoSuchProcess:
             pass
 
     # Kill shell itself.
-    p.kill()
-
+    p.terminate()
+    try:
+        p.wait(timeout=GRACEFUL_TERMINATION_TIME_S)
+    except psutil.TimeoutExpired:
+        p.kill()
```

Comment:
this `p.kill()` might raise an `psutil.NoSuchProcess` as well

Reply:
Good catch!